wagtail/models/sites.py

```diff
@@ -209,42 +212,58 @@ def get_site_root_paths():
-                            SiteRootPath(
-                                site.id,
-                                root_page.url_path,
-                                site.root_url,
-                                root_page.locale.language_code,
-                            )
-                            for root_page in site.root_page.get_translations(
-                                inclusive=True
-                            ).select_related("locale")
-                        ]
-                    )
-                else:
-                    result.append(
+
+        # Reuse previously generated value if the method was already called
+        # for the current thread
+        result = getattr(_root_paths_cache, "value", None)
+        if result is not None:
+            # Return a copy to prevent accidental pollution of cached values
+            return deepcopy(result)
+
+        result = []
+        for site in Site.objects.select_related(
+            "root_page", "root_page__locale"
+        ).order_by("-root_page__url_path", "-is_default_site", "hostname"):
+            if getattr(settings, "WAGTAIL_I18N_ENABLED", False):
+                result.extend(
+                    [
                         SiteRootPath(
                             site.id,
-                            site.root_page.url_path,
+                            root_page.url_path,
                             site.root_url,
-                            site.root_page.locale.language_code,
+                            root_page.locale.language_code,
                         )
+                        for root_page in site.root_page.get_translations(
+                            inclusive=True
+                        ).select_related("locale")
+                    ]
+                )
+            else:
+                result.append(
+                    SiteRootPath(
+                        site.id,
+                        site.root_page.url_path,
+                        site.root_url,
+                        site.root_page.locale.language_code,
                     )
+                )
 
-            cache.set("wagtail_site_root_paths", result, 3600)
+        # Cache result for the current thread
+        _root_paths_cache.value = result
 
         return result
+
+    @staticmethod
+    def clear_site_root_paths_cache() -> None:
+        """
+        As a courtesy, we clear the cache when certain actions might affect
+        site root path values. However, because we use an in-memory cache of
+        sorts, we can only ever do this for the same running instance.
+
+        While this does mean that other running instances of the project may
+        have active threads using invalid data, we know that the life of those
+        threads is typically short, so can live with it. The data is typically
+        cached for a thread's corresponding ``HttpRequest`` anyway.
+        """
+        global _root_paths_cache
+        _root_paths_cache = Local()
```

Comment:
I was in two minds here about whether to trash the entire thing here, or just unset the value for the current thread only... Which felt 'fairer' from a 'multiple running instances' perspective, and probably has slightly less potential to cause issues should multiple threads be trying to access values concurrently. Thoughts?